### ISSUE TITLE
Handlers from dependencies are not deduplicated in ansible-core 2.15.0

### ISSUE THREAD

**Issue by NeonSludge** [ansible/ansible#81013]
### Summary

Handlers included from other roles via role dependencies are not deduplicated properly in ansible-core 2.15.0. This behaviour is specific to 2.15.0 and doesn't seem to be documented anywhere.

### Issue Type

Bug Report

### Component Name

ansible-core

### Ansible Version

```console
$ ansible --version
ansible [core 2.15.0]
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /tmp/ansible-test/lib64/python3.11/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /tmp/ansible-test/bin/ansible
  python version = 3.11.2 (main, Apr  5 2023, 11:57:00) [GCC 8.5.0 20210514 (Red Hat 8.5.0-18)] (/tmp/ansible-test/bin/python3.11)
  jinja version = 3.1.2
  libyaml = True
```

### Configuration

```console
$ ansible-config dump --only-changed -t all
CONFIG_FILE() = None
```

### OS / Environment

AlmaLinux 8

### Steps to Reproduce

1. Create a "global" role with a handler:

```
# roles/global/handlers/main.yml
- name: a global handler
  debug:
    msg: "a global handler has been triggered"
  listen: "global handler"
```

2. Create two roles with tasks that notify this handler, define the "global" role as a dependency in both roles:

```
# roles/role1/meta/main.yml
dependencies:
  - global
```

```
# roles/role1/tasks/main.yml
- name: role1/task1
  debug:
  changed_when: true
  notify: "global handler"
```

```
# roles/role2/meta/main.yml
dependencies:
  - global
```

```
# roles/role2/tasks/main.yml
- name: role2/task1
  debug:
  changed_when: true
  notify: "global handler"
```

3. Create a playbook:

```
# playbook.yml
- hosts: localhost
  roles:
    - role1
    - role2
```

4. Resulting file tree:
```
.
├── playbook.yml
└── roles
    ├── global
    │   └── handlers
    │       └── main.yml
    ├── role1
    │   ├── meta
    │   │   └── main.yml
    │   └── tasks
    │       └── main.yml
    └── role2
        ├── meta
        │   └── main.yml
        └── tasks
            └── main.yml
```

5. Run the playbook with Ansible 2.15, verify that the handler has been invoked twice.

### Expected Results

Ansible 2.14 deduplicates the handler:

```
$ ansible-playbook --version
ansible-playbook [core 2.14.6]
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /tmp/ansible-test/lib64/python3.11/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /tmp/ansible-test/bin/ansible-playbook
  python version = 3.11.2 (main, Apr  5 2023, 11:57:00) [GCC 8.5.0 20210514 (Red Hat 8.5.0-18)] (/tmp/ansible-test/bin/python3.11)
  jinja version = 3.1.2
  libyaml = True
```

```
$ ansible-playbook playbook.yml
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] ****************************************************************************************************************************************************************************

TASK [Gathering Facts] **********************************************************************************************************************************************************************
ok: [localhost]

TASK [role1 : role1/task1] ******************************************************************************************************************************************************************
changed: [localhost] => {
    "msg": "Hello world!"
}

TASK [role2 : role2/task1] ******************************************************************************************************************************************************************
changed: [localhost] => {
    "msg": "Hello world!"
}

RUNNING HANDLER [global : a global handler] *************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "a global handler has been triggered"
}

PLAY RECAP **********************************************************************************************************************************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

```

Ansible 2.15 runs this handler twice (see actual results).

### Actual Results

```console
$ ansible-playbook playbook.yml
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] ****************************************************************************************************************************************************************************

TASK [Gathering Facts] **********************************************************************************************************************************************************************
ok: [localhost]

TASK [role1 : role1/task1] ******************************************************************************************************************************************************************
changed: [localhost] => {
    "msg": "Hello world!"
}

TASK [role2 : role2/task1] ******************************************************************************************************************************************************************
changed: [localhost] => {
    "msg": "Hello world!"
}

RUNNING HANDLER [global : a global handler] *************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "a global handler has been triggered"
}

RUNNING HANDLER [global : a global handler] *************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "a global handler has been triggered"
}

PLAY RECAP **********************************************************************************************************************************************************************************
localhost                  : ok=5    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/devel/ISSUE_HELP.md)

**Comment by mkrizek** [comment]
The change in behavior was introduced in #78661.

The cause appears to be that the `global` role is created each time it is used as a dependency. In effect the handler is added to handlers twice as well. Since the `listen` topic is used as the notification both handlers are notified and executed.

I don't know what the actual playbook in use is but if possible notifying by a handler name (`a global handler`) would result in the desired behavior.

I can't comment on what the resolution of this issue will be.

---
Also to clarify, handlers are never deduplicated. They are added to a global, play scoped, list and when notified the last one added is executed.

**Comment by egor-tensin** [comment]
Thank god I wasn't the only one seeing this; see for example this GitHub Actions [run](https://github.com/egor-tensin/maintenance/actions/runs/5661637555/job/15339957989#step:6:256), where my laptop was rebooted about 5 times instead of once. The playbook is [here](https://github.com/egor-tensin/maintenance/tree/587e746ed56bf8e99a13e8ba237e84a71ec98d49), the handler is defined [here](https://github.com/egor-tensin/maintenance/blob/587e746ed56bf8e99a13e8ba237e84a71ec98d49/roles/common/handlers/main.yml), and is included as a role dependency in a bunch of places: [1](https://github.com/egor-tensin/maintenance/blob/587e746ed56bf8e99a13e8ba237e84a71ec98d49/roles/pacman/meta/main.yml#L2), [2](https://github.com/egor-tensin/maintenance/blob/587e746ed56bf8e99a13e8ba237e84a71ec98d49/roles/apt/meta/main.yml#L2), [3](https://github.com/egor-tensin/maintenance/blob/587e746ed56bf8e99a13e8ba237e84a71ec98d49/roles/yay/meta/main.yml#L2), etc.

**Comment by mkrizek** [comment]
@NeonSludge @egor-tensin there is a PR with proposed fix in https://github.com/ansible/ansible/pull/81358 in case you are interested in testing.

**Comment by egor-tensin** [comment]
Seems to have fixed this particular issue for me. Test playbook is [test_playbook.zip]. Ansible 2.15.2 executes the handler twice:

```
PLAY [Maintenance] *******************************************************************

TASK [Gathering Facts] ***************************************************************
ok: [localhost]

TASK [tasks : Test task] *************************************************************
changed: [localhost] => {
    "msg": "Test task"
}

TASK [tasks : Flush handlers] ********************************************************

RUNNING HANDLER [handlers : Test handler] ********************************************
ok: [localhost] => {
    "msg": "Test handler"
}

RUNNING HANDLER [handlers : Test handler] ********************************************
ok: [localhost] => {
    "msg": "Test handler"
}

PLAY RECAP ***************************************************************************
localhost                  : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

The version from the PR executes it only once, as expected:

```
PLAY [Maintenance] ********************************************************************

TASK [Gathering Facts] ****************************************************************
ok: [localhost]

TASK [tasks : Test task] **************************************************************
changed: [localhost] => {
    "msg": "Test task"
}

TASK [tasks : Flush handlers] *********************************************************

RUNNING HANDLER [handlers : Test handler] *********************************************
ok: [localhost] => {
    "msg": "Test handler"
}

PLAY RECAP ****************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

[test_playbook.zip](https://github.com/ansible/ansible/files/12207703/test_playbook.zip)

However, I do observe another issue when handlers get skipped due to a wrong `false_condition` - not the one specified by the handler. Trying to make a minimal reproducible example now.

**Comment by egor-tensin** [comment]
I reproduced another issue I was having with the PR version. The test playbook is in [test_playbook2.zip](https://github.com/ansible/ansible/files/12207720/test_playbook2.zip).

Role `tasks2` has an uncoditional dependency on the `handlers` role (which defines the handlers), and a _conditional_ dependency on the `tasks1` role (with the condition being set to `false`). Ansible 2.15.2 executes the handler multiple times, but it _does_ execute it:

```
PLAY [Maintenance] *******************************************************************

TASK [Gathering Facts] ***************************************************************
ok: [localhost]

TASK [handlers : Set a test fact] ****************************************************
ok: [localhost]

TASK [tasks1 : Test task 1] **********************************************************
skipping: [localhost]

TASK [tasks1 : Flush handlers] *******************************************************
skipping: [localhost]

TASK [tasks2 : Test task 2] **********************************************************
changed: [localhost] => {
    "msg": "Test task 2"
}

TASK [tasks2 : Flush handlers] *******************************************************

RUNNING HANDLER [handlers : Test handler] ********************************************
ok: [localhost] => {
    "msg": "Test handler"
}

RUNNING HANDLER [handlers : Test handler] ********************************************
ok: [localhost] => {
    "msg": "Test handler"
}

RUNNING HANDLER [handlers : Test handler] ********************************************
skipping: [localhost]

PLAY RECAP ***************************************************************************
localhost                  : ok=5    changed=1    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0
```

The PR version doesn't execute the handler at all:

```
PLAY [Maintenance] ********************************************************************

TASK [Gathering Facts] ****************************************************************
ok: [localhost]

TASK [handlers : Set a test fact] *****************************************************
ok: [localhost]

TASK [tasks1 : Test task 1] ***********************************************************
skipping: [localhost]

TASK [tasks1 : Flush handlers] ********************************************************
skipping: [localhost]

TASK [tasks2 : Test task 2] ***********************************************************
changed: [localhost] => {
    "msg": "Test task 2"
}

TASK [tasks2 : Flush handlers] ********************************************************

RUNNING HANDLER [handlers : Test handler] *********************************************
skipping: [localhost]

PLAY RECAP ****************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0
```

Running with `-vvv` shows that Ansible assigns the condition to the executed handler, even though it's also included unconditionally:

```
TASK [tasks2 : Test task 2] ***********************************************************
task path: /home/egor/workspace/personal/test_playbook2/roles/tasks2/tasks/main.yml:1
Notification for handler test_handler has been saved.
changed: [localhost] => {
    "msg": "Test task 2"
}

TASK [tasks2 : Flush handlers] ********************************************************
task path: /home/egor/workspace/personal/test_playbook2/roles/tasks2/tasks/main.yml:7
NOTIFIED HANDLER handlers : Test handler for localhost
META: triggered running handlers for localhost

RUNNING HANDLER [handlers : Test handler] *********************************************
task path: /home/egor/workspace/personal/test_playbook2/roles/handlers/handlers/main.yml:1
skipping: [localhost] => {
    "false_condition": "test_fact"
}
```

Looks wrong to me.

**Comment by mkrizek** [comment]
> I reproduced another issue I was having with the PR version. The test playbook is in [test_playbook2.zip](https://github.com/ansible/ansible/files/12207720/test_playbook2.zip).

Please do not bundle reproducers in an archive, rather copy&paste it directly or use one of the paste services so we can inspect the playbook without downloading unknown archives.

Copy&pasting the reproducer:

`~ % cat Downloads/test_playbook2/test.yml`
```yaml
- name: Maintenance
  hosts: [localhost]
  roles:
    - handlers
    - tasks2
```
`~ % cat Downloads/test_playbook2/roles/handlers/tasks/main.yml`
```yaml
- name: Set a test fact
  ansible.builtin.set_fact: 'test_fact=false'
```
`~ % cat Downloads/test_playbook2/roles/handlers/handlers/main.yml`
```yaml
- name: Test handler
  ansible.builtin.debug:
    msg: Test handler
  listen: test_handler
```
`~ % cat Downloads/test_playbook2/roles/tasks2/meta/main.yml`
```yaml
dependencies:
  - role: handlers
  - {role: tasks1, when: test_fact}
```
`~ % cat Downloads/test_playbook2/roles/tasks2/tasks/main.yml`
```yaml
- name: Test task 2
  ansible.builtin.debug:
    msg: Test task 2
  changed_when: true
  notify: test_handler

- name: Flush handlers
  ansible.builtin.meta: flush_handlers
```
`~ % cat Downloads/test_playbook2/roles/tasks1/meta/main.yml`
```yaml
dependencies:
  - role: handlers
```
`~ % cat Downloads/test_playbook2/roles/tasks1/tasks/main.yml`
```yaml
- name: Test task 1
  ansible.builtin.debug:
    msg: Test task 1
  changed_when: true
  notify: test_handler

- name: Flush handlers
  ansible.builtin.meta: flush_handlers
```

To me this is intended behavior. Note that all (so including ones from roles) handlers are added to the global scope, handlers with the same name overriding ones added previously.

In the reproducer above the `Test handler` handler from the `handler` role is added three times.
1. as part of the `handlers` role directly from `- roles:`
2. as part of the `handlers` role as a dependency of `tasks2` role
3. as part of the `handlers` role as a dependency of `tasks1` role as a dependency of `tasks2` role

The third instance is added last and it is this instance that is being used. The reason why it is skipped is because the `tasks1` dependency has a conditional (evaluated to `False`) attached that the handler inherits. See https://github.com/ansible/ansible/issues/80925.

**Comment by egor-tensin** [comment]
It worked fine on every version before 2.15. Was it not intended behaviour then?

**Comment by egor-tensin** [comment]
This is pretty counter-intuitive honestly, if you ask me. So the workaround is to explicitly order the depedencies, hoping that you don't forget to put an unconditional dependency last? To be honest, this doesn't seem sustainable in the long run, but I can probably live with this.

**Comment by mkrizek** [comment]
However, from https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_handlers.html#naming-handlers:

> Alternately handlers sharing the same name can all be notified and executed if they listen on the same topic by notifying that topic.

Which brings me back to https://github.com/ansible/ansible/issues/81013#issuecomment-1584573232 as the current behavior works as documented?

We need to decide which behavior is "correct".

**Comment by bcoca** [comment]
The behavior of 'last defined wins' is what is documented and expected, handlers are always imported once the role is, conditionals apply to task execution (handlers are still tasks) not to handler import.

**Comment by egor-tensin** [comment]
Let me make up a scenario: a playbook author imports roles `handlers` and `tasks` that he doesn't have deep knowledge about (maybe he's not the author, or he simply forgot). He doesn't even know that `tasks` imports `handlers` itself. Wouldn't it be better to allow them _not_ have deep knowledge about what roles import what? Because if they don't know that `tasks` imports `handlers` (which they shouldn't?), they could import `tasks` conditionally, which breaks the handlers completely.

Anyway, you seem to be pretty set on this, so I guess I'm done :) Thanks for all the work!

**Comment by bcoca** [comment]
tasks don't import handlers, importing a role includes 'handlers and vars' that are part of the role. If you JUST want tasks, use `import_tasks`.
